Add URL Protocol Handler article

tomayac opened the pull request:
Fixes #4481

Changes proposed in this pull request:

- Add the article

jpmedley commented:
@tomayac Is this ready for review?

tomayac commented:
> @tomayac Is this ready for review?

Not yet. I ping you once it's ready.  Thanks for asking though!

tomayac commented:
@jpmedley, this article is now ready for review. Thanks in advance!

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+
+## How to use URL protocol handler registration for PWAs
+
+The API of the URL protocol handler registration for PWAs feature is modeled closely along the
+existing API of `navigator.registerProtocolHandler()`. Just this time the information is passed
```
```suggestion
The API for URL protocol handler registration is modeled closely on  
`navigator.registerProtocolHandler()`. Just this time the information is passed
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+
+The API of the URL protocol handler registration for PWAs feature is modeled closely along the
+existing API of `navigator.registerProtocolHandler()`. Just this time the information is passed
+declaratively via the Web Application Manifest in a new property `"protocol_handlers"` that takes an
```
```suggestion
declaratively via the Web Application Manifest in a new property called `"protocol_handlers"` that takes an
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+existing API of `navigator.registerProtocolHandler()`. Just this time the information is passed
+declaratively via the Web Application Manifest in a new property `"protocol_handlers"` that takes an
+array of objects with the two required keys `"protocol"` and `"url"`. The code snippet below shows
+the relevant excerpt of a Web Application Manifest that would register the protocols `web+tea` and
```
```suggestion
how to register `web+tea` and
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+array of objects with the two required keys `"protocol"` and `"url"`. The code snippet below shows
+the relevant excerpt of a Web Application Manifest that would register the protocols `web+tea` and
+`web+coffee`. The values are strings containing the URL of the handler with the required `%s`
+placeholder that will be replaced with the escaped URL to be handled.
```
```suggestion
placeholder for the escaped URL.
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+Handler registrations are synchronized with the latest manifest version provided by the app. There
+are two cases:
+
+1. An update that adds new handlers triggers handler registration (separate from app installation).
+1. An update that removes handlers triggers handler unregistration (separate from app
+   uninstallation).
```
```suggestion
* An update that adds new handlers triggers handler registration (separate from app installation).
* An update that removes handlers triggers handler unregistration (separate from app
   uninstallation).
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+   [https://protocol-handler-link.glitch.me/](https://protocol-handler-link.glitch.me/). This will
+   open a new browser tab with three links. Click on the first or the second (latte macchiato or
+   americano). The browser will now show you a prompt and ask if you are fine with the app being a
+   protocol handler for the `web+coffee` protocol. If you agree, the PWA will open and show the
```
We're really asking users to judge whether schemes can be used?

jpmedley replied:
Or is it presented in a non-technical way?

tomayac replied:
This is what it looks like:

<img width="313" alt="Screen Shot 2021-05-06 at 10 00 47" src="https://user-images.githubusercontent.com/145676/117262839-f6bb8300-ae51-11eb-8164-9fb2a007de8e.png">

Note that the pitch for the feature the article is about is that there would be _no_ prompt, but the installation per se would be seen as the user being fine with the registration happening.

tomayac replied:
(To clarify: the image above is for the JS-based registration.)

fabiorocha replied:
That's correct, this is for JS-based registration but we will have a similar experience on the first time a protocol handler for a web app is invoked. This was a requirement from security/privacy folks.

![image](https://user-images.githubusercontent.com/1686833/117515279-a116d180-af4a-11eb-9272-40b9e41dd23c.png)

fabiorocha replied:
[Edit 1]: I now see this is also below. Apologies. I'm reviewing this as I read it :-)

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+
+## Security considerations
+
+Since PWA installation requires the context to be secure, the present feature inherits this
```
```suggestion
Since PWA installation requires the context to be secure, protocol handling inherits this
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+### Allowlist of protocols
+
+Just like with `registerProtocolHandler()` there is an allowlist of protocols that apps can register
+to become the handler for.
```
```suggestion
to handle.
```

jpmedley commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -48,49 +48,220 @@ The secure-content-only `Navigator` method
+
+On the first launch of the PWA due to an invoked protocol, the user will be presented with a
+permission dialog. This dialog will display the app name and origin of the app, and ask the user if
+the app is allowed to handle links from the protocol. If a user rejects the permission dialog, the
```
Again, how is this presented? If I'm usability minded, I might be disinclined to use this.

tomayac replied:
It shows a prompt if the PWA should be used (every time) to handle links of the given protocol. I can't seem to reset the state after saying "every time", so can't show a screenshot.

jpmedley replied:
I'm satisfied now that this is fine. Would a web developer have the same concern as me? Would it be worth mentioning in passing that this will be presented in a user-friendly way? I totally admit this is a judgement call.

tomayac replied:
To be quite honest I don't think regular users will understand what this is about (by the pure nature of the thing). But then on the other side, platform-specific apps like YouTube (the Android app) register themselves as handlers, so users are used to this kind of behavior where an app launches when a link is clicked (despite knowing that youtube.com as a website exists in a parallel world).

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -1,28 +1,28 @@
 ---
 layout: post
-title: URL protocol handler registration for PWAs
+title: URL protocol handler registration for installed PWAs
```
It only works for installed PWAs, so I thought mentioning it right in the heading was the right thing to do.

jpmedley replied:
OK... but nothing works for _uninstalled_ PWAs, so I still don't see what this adds.

tomayac replied:
Let me remove the word "installed" then.

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+---
+layout: post
+title: URL protocol handler registration for installed PWAs
```
```suggestion
title: URL protocol handler registration for PWAs
```

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+The present _URL protocol handler registration for PWAs_ proposal is about offering protocol handler
+registration as part of a PWA installation through its manifest. After registering a PWA as a
+protocol handler, when a user clicks on a hyperlink with a specific scheme such as `mailto` ,
+`ms-word`, or `web+music` from a browser or a platform-specific app, the registered PWA will open
```
```suggestion
`bitcoin`, or `web+music` from a browser or a platform-specific app, the registered PWA will open
```

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+- Similarities include requirements around the list of schemes allowed to be registered, and
+  the name and format of parameters, etc.
+- Differences in the manifest-based registration are subtle, but might be useful to enhance the
+  experience for PWA users. For example, manifest-based PWA registration may not require an
```
For security/privacy reasons we will always require additional user action, but it will be when invoking a link that is handled by an app. It will be in the form of a permission dialog (i.e.: allow or cancel) or a disambiguation dialog (i.e.: the open with dialog on Windows). Not sure if you want to talk about those cases in this document but it might be useful. If so we could help with screenshots and etc.

fabiorocha replied:
[edit 1]: Ah, I see you talk about this further below.

tomayac replied:
The Explainer is structured similarly.

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+### Use cases
+
+- In a word processing PWA, the user in a document encounters a link to a presentation like
+  `ms-powerpoint://deck2378465`. When the user clicks on the link, the presentation PWA
```
Another protocol that hasn't made into the safelist yet. Maybe replace with the coffee/tea example you have? Or just do web+powerpoint?

tomayac replied:
I'll use a neutral presentation software term.

tomayac replied:
(Again this comes straight from the Explainer, similar to above, you might want to update accordingly.)

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+| Step                                     | Status                   |
+| ---------------------------------------- | ------------------------ |
+| 1. Create explainer                      | [Complete][explainer]    |
+| 2. Create initial draft of specification | Not started              |
```
```suggestion
| 2. Create initial draft of specification | [In Progress][specification]              |
```

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+
+[explainer]:
+  https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md
+[github]: https://github.com/MicrosoftEdge/MSEdgeExplainers/tree/main/URLProtocolHandler
```
```suggestion
[specification]: https://github.com/w3c/manifest/pull/972
[github]: https://github.com/MicrosoftEdge/MSEdgeExplainers/tree/main/URLProtocolHandler
```

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+
+## Useful links
+
+- [Explainer][explainer]
```
```suggestion
- [Explainer][explainer]
- [Spec proposal][specification]
```

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+
+You can see a demo of URL protocol handler registration for PWAs on Glitch.
+
+1. Go to [https://protocol-handler.glitch.me/](https://protocol-handler.glitch.me/), install the
```
Not sure if we want to remind readers that step 0 is enabling the flag.

tomayac replied:
I'll add it here and to the demo.

fabiorocha commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+On the first launch of the PWA due to an invoked protocol, the user will be presented with a
+permission dialog. This dialog will display the app name and origin of the app, and ask the user if
+the app is allowed to handle links from the protocol. If a user rejects the permission dialog, the
+protocol handler is unregistered with the operating system.
```
This is outdated. The only way to unregister protocols for web apps is by uninstalling the app. Cancelling the consent prompt merely closes it. The user consent will be saved in prefs if they click allow.

![image](https://user-images.githubusercontent.com/1686833/117515587-8e50cc80-af4b-11eb-905a-78ef7f9a87fb.png)

tomayac replied:
I have updated this accordingly.

fabiorocha reviewed (changes requested):
Thanks for putting this together Thomas. I left a few comments/suggestions.

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+date: 2021-05-05
+description: |
+  After registering a PWA as a protocol handler, when a user clicks on a hyperlink with a specific
+  scheme such as mailto, ms-word, or web+music from a browser or a platform-specific app,
```
I was wondering about this, but since it was mentioned in the Explainer, I used it in the article. You might want to update the Explainer accordingly then (or add a remark that this is a potential future addition).

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+### Use cases
+
+- In a word processing PWA, the user in a document encounters a link to a presentation like
+  `ms-powerpoint://deck2378465`. When the user clicks on the link, the presentation PWA
```
```suggestion
  `web+presentations://deck2378465`. When the user clicks on the link, the presentation PWA
```

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+
+You can see a demo of URL protocol handler registration for PWAs on Glitch.
+
+1. Go to [https://protocol-handler.glitch.me/](https://protocol-handler.glitch.me/), install the
```
```suggestion
1. If you have not done so before, [enable the flag](#enabling-via-chrome:flags).
1. Go to [https://protocol-handler.glitch.me/](https://protocol-handler.glitch.me/), install the
```

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,267 @@
+
+On the first launch of the PWA due to an invoked protocol, the user will be presented with a
+permission dialog. This dialog will display the app name and origin of the app, and ask the user if
+the app is allowed to handle links from the protocol. If a user rejects the permission dialog, the
+protocol handler is unregistered with the operating system.
```
```suggestion
the app is allowed to handle links from the protocol. If a user rejects the permission dialog, the
registered protocol handler will be ignored by the operating system.
To unregister the protocol handler, the user needs to uninstall the PWA that registered it.
```

tomayac commented:
> Thanks for putting this together Thomas. I left a few comments/suggestions.

Thanks, @fabiorocha, very helpful. I have addressed all your comments. PTAL.

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,272 @@
+[Diego González](https://www.linkedin.com/in/diekus/),
+[Connor Moody](https://www.linkedin.com/in/connor-d-moody/), and
+[Samuel Tang](https://www.linkedin.com/in/tangsamuel/) from the Microsoft Edge team. This article
+was reviewed by [Joe Medley](https://github.com/jpmedley). Hero image by
```
```suggestion
was reviewed by [Joe Medley](https://github.com/jpmedley) and Fabio Rocha. Hero image by
```

tomayac commented on src/site/content/en/blog/url-protocol-handler/index.md:
```diff
@@ -0,0 +1,271 @@
+  Let installed PWAs handle links that use a specific protocol for a more integrated experience.
+authors:
+  - thomassteiner
+date: 2021-05-10
```
```suggestion
date: 2021-05-11
```

tomayac commented:
Thanks everyone! This will go live soon. Excited to see this land!
